packages/caliper-ethereum/lib/ethereum.js

```diff
@@ -0,0 +1,224 @@
+     * @param {Number} timeout Request timeout, in seconds.
+     * @return {Promise<TxStatus>} Result and stats of the transaction invocation.
+     */
+    async sendTransaction(context, contractID, contractVer, methodCall, timeout) {
+        let status = new TxStatus();
+        try {
+            context.engine.submitCallback(1);
+            let receipt = null;
+            if (methodCall.args) {
+                receipt = await context.contracts[contractID].methods[methodCall.verb](...methodCall.args).send({from: context.fromAddress});
+            } else {
+                receipt = await context.contracts[contractID].methods[methodCall.verb]().send({from: context.fromAddress});
+            }
+            status.SetID(receipt.transactionHash);
+            status.SetResult(receipt);
+            status.SetVerification(true);
+            status.SetStatusSuccess();
+        } catch (err) {
+            status.SetStatusFail();
+            logger.error('Failed tx on ' + contractID + ' calling method ' + methodCall.verb);
+        }
+        return Promise.resolve(status);
+    }
+
+    /**
+     * Query the given smart contract according to the specified options.
+     * @param {object} context The Ethereum context returned by {getContext}.
+     * @param {string} contractID The name of the contract.
+     * @param {string} contractVer The version of the contract.
+     * @param {string} key The argument to pass to the smart contract query.
+     * @param {string} [fcn=query] The contract query function name.
+     * @return {Promise<object>} The promise for the result of the execution.
+     */
+    async queryState(context, contractID, contractVer, key, fcn = 'query') {
+        let status = new TxStatus();
+
+        if (context.engine) {
+            context.engine.submitCallback(1);
+        }
+
+        try {
+            let receipt = await context.contracts[contractID].methods[fcn](key).call();
+            status.SetID(null);
+            status.SetResult(receipt);
+            status.SetVerification(true);
+            status.SetStatusSuccess();
+        } catch (err) {
+            status.SetStatusFail();
+            logger.error('Failed querying state on ' + contractID + ' calling method ' + fcn  + ' for key ' + key);
+        }
+        return Promise.resolve(status);
+    }
+
+    /**
+     * Fetch the address for the contract with the given label from the registry
+     * @param {string} contract_id id to search for
+     * @return {string} The contract address
+     */
+    lookupContract(contract_id) {
+        return this.registry.methods.lookup(contract_id).call();
+    }
+
+    /**
+     * Binds the address to the label registering on the registry
+     * @param {string} contract_id contract_id to bind address to
+     * @param {string} address deployed contract address
+     * @returns {Promise<PromiEvent>} PromiEvent is a web3js type. Refer to web3js doc.
+     */
+    bindContract(contract_id, address) {
```

Comment:
I think keeping the contract `name => address` mapping on-chain is a bit much. There's a "Caliper-way" of passing this information from the master process to the worker processes.
There's a `prepareClients` function on the `BlockchainInterface`, which is called in the master process, and in it, you can build objects to pass to every worker client:
https://github.com/hyperledger/caliper/blob/6ddd6003da8772e9065cf5f179b177e8b47705d9/packages/caliper-core/lib/blockchain-interface.js#L48-L53

By default, it sends an empty object, but you can pass the name=>address mapping for all clients this way. On the worker client side, you'll get this object in `getContext`, in the `args` argument.

This way you can reduce the complexity (and necessary artifacts) for your adapter.

Reply:
Now result holds contract abi and address. Each client is able to construct its own Contract instance.